pkg/unixserver/server.go

```diff
@@ -257,16 +275,31 @@ func (a *CDMetrics) Describe(ch chan<- *prometheus.Desc) {
 	}
 }
 
-//Collect implements prometheus.Collector.
+//Collect implements prometheus.Collector
 func (a *CDMetrics) Collect(ch chan<- prometheus.Metric) {
-	for _, metric := range a.metrics {
-		for _, labeled_metric := range metric {
-			ch <- prometheus.NewMetricWithTimestamp(labeled_metric.timeStamp, prometheus.MustNewConstMetric(labeled_metric.metricDesc, labeled_metric.valueType, labeled_metric.metric,
-				labeled_metric.host, labeled_metric.pluginInstance, labeled_metric.typeInstance))
+	for metricName, metric := range a.metrics {
+		for labelName, labeledMetric := range metric {
+
+			//Delete expired label metrics
+			if labeledMetric.expired() {
```

Comment:
I think this code should move to after the NewMetricWithTimestamp() code (L290), that way the scrape will still see the last value before it gets dropped. 

AFAICT this new timeout code is only running in response to scrapes of the prometheus endpoint. This seems sub-optimal because if Prometheus dies we could possibly accumulate a large number of dead metrics, but in practice this is probably not a huge problem. With the code as it is, though, when prometheus returns, I think it will not even get the last sample of now-dead metrics that it has missed. I feel like it would be better to let them out, since they are timestamped anyways, it's better than just dropping them sight unseen.

Open to clarifications or counter-arguments, please!

Reply:
If Promethues dies (correct me if I am wrong), new values for the same metric will overwrite the old values and you will end up with no more metric storage than in the regular case where Prom is running.  I would argue for dropping before if the 5x interval has expired.  If promethues is down for a day, a value older than a day would not (IMO) have any value.  Data's value eponentially decays with time. :).

Reply:
Yes, new values for the same metric will overwrite the old ones; I was referring to different metrics that may have come and gone while prometheus was down. In either case I agree that metrics are less valuable the older they are and this isn't a huge problem. I still think moving the code to the other side of NewMetricWithTimestamp() is better, but I haven't tested it.

Reply:
I moved responsibility of expiring metrics to a separate goroutine (see below), so this is no longer an issue.